src/adapters/v01/listings_adapter.js

```diff
@@ -0,0 +1,372 @@
+      )
+    }
+    const purchaseType = ipfsData.purchaseType || unitListingType
+    let validateFn, schema
+    if (purchaseType === unitListingType) {
+      validateFn = validateUnitPurchase
+      schema = unitPurchaseSchema
+    } else if (purchaseType === fractionalListingType) {
+      validateFn = validateFractionalPurchase
+      schema = fractionalPurchaseSchema
+    }
+    validate(validateFn, schema, ipfsData)
+
+    const ipfsHash = await this.ipfsService.submitFile(ipfsData)
+    const ipfsBytes32 = this.contractService.getBytes32FromIpfsHash(ipfsHash)
+    return await this.contractService.call(
+      listingsContract,
+      'requestPurchase',
+      [listingIndex, ipfsBytes32],
+      { value: offerWei, gas: 350000 }
+    )
+  }
+
+  async acceptPurchaseRequest(
+    listingIndex,
+    purchaseIndex,
+    ipfsData,
+    confirmationCallback
+  ) {
+    const ipfsHash = await this.ipfsService.submitFile(ipfsData)
+    const ipfsBytes32 = this.contractService.getBytes32FromIpfsHash(ipfsHash)
+    return await this.contractService.call(
+      listingsContract,
+      'acceptPurchaseRequest',
+      [listingIndex, purchaseIndex, ipfsBytes32],
+      {},
+      confirmationCallback
+    )
+  }
+
+  async rejectPurchaseRequest(
+    listingIndex,
+    purchaseIndex,
+    ipfsData,
+    confirmationCallback
+  ) {
+    const ipfsHash = await this.ipfsService.submitFile(ipfsData)
+    const ipfsBytes32 = this.contractService.getBytes32FromIpfsHash(ipfsHash)
+    return await this.contractService.call(
+      listingsContract,
+      'rejectPurchaseRequest',
+      [listingIndex, purchaseIndex, ipfsBytes32],
+      {},
+      confirmationCallback
+    )
+  }
+
+  async buyerFinalizePurchase(
+    listingIndex,
+    purchaseIndex,
+    ipfsData,
+    confirmationCallback
+  ) {
+    validate(validateReview, reviewSchema, ipfsData)
+    const ipfsHash = await this.ipfsService.submitFile(ipfsData)
+    const ipfsBytes32 = this.contractService.getBytes32FromIpfsHash(ipfsHash)
+    return await this.contractService.call(
+      listingsContract,
+      'buyerFinalizePurchase',
+      [listingIndex, purchaseIndex, ipfsBytes32],
+      {},
+      confirmationCallback
+    )
+  }
+
+  async sellerFinalizePurchase(
+    listingIndex,
+    purchaseIndex,
+    ipfsData,
+    confirmationCallback
+  ) {
+    validate(validateReview, reviewSchema, ipfsData)
+    const ipfsHash = await this.ipfsService.submitFile(ipfsData)
+    const ipfsBytes32 = this.contractService.getBytes32FromIpfsHash(ipfsHash)
+    return await this.contractService.call(
+      listingsContract,
+      'sellerFinalizePurchase',
+      [listingIndex, purchaseIndex, ipfsBytes32],
+      {},
+      confirmationCallback
+    )
+  }
+
+  async getPurchase(listingIndex, purchaseIndex) {
+    return await getPurchaseHelper(
+      this.contractService,
+      this.ipfsService,
+      listingIndex,
+      purchaseIndex
+    )
+  }
+
+  async getPurchases(listingIndex) {
```

Comment:
Maybe add a TODO about this not scaling once there are many purchases because of memory usage? I'm not sure about the memory limits of in-browser JS, but there's probably some practical limit.